lastSequenceNr should reflect the snapshot sequence and not start with 0 when journal is empty

erinmez opened the pull request:
<!--
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [-] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?
-->
disclaimer: I'm really new to the internals, so I'm happy to learn and improve this PR :-)
## Purpose
When a journal log is non-existent but a snapshot can be found (let's say with sequence 42), the state is correctly offered, however, the lastSequenceNr (which contains 42)  is overwritten with the highest sequence number that was found during journal recovery (which is 0). When a snapshot is taken now, its snapshot sequence will be most probably lower (let's say 10).  When the actor recovers anew, the snapshot with sequence 42 will be found, hiding the more recent changes in the snapshot with 10.
## Changes

- My proposed change is to use either the lastSequenceNr (which is already set to the value from the snapshot) or the one found from the journal, whichever is higher (see the change in Eventsourced).

- SnapshotRecoveryWithEmptyJournalSpec simulates the described problem above

## Background Context
In one of our projects we can live with a transient but fast journal. That led us to a really simple k8s pod setup with 1 jvm, running 1 actor system, LeveldbJournal, LocalSnapshotStore & no akka cluster needed.  We store the journal on the local node but keep snapshots on a persistent volume. The pod is rescheduled to a different node from time to time. Because we leave the journal on local storage, the journal will be empty when the pod is restarted on the new node (which is fine and expected in our case).

erinmez committed:
lastSequenceNr should reflect the snapshot sequence number and not start
at 0 when there was no journal to begin with.

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

patriknw commented on akka-persistence/src/test/scala/akka/persistence/SnapshotRecoveryWithEmptyJournalSpec.scala:
```diff
@@ -0,0 +1,91 @@
+}
+
+class SnapshotRecoveryWithEmptyJournalSpec extends PersistenceSpec(PersistenceSpec.config(
+  "inmem", // with leveldb the error cannot be reproduced (counterKey is not reset after deleteMessages() )
```
The change shouldn't be needed, because according to the journal protocol `asyncReadHighestSequenceNr` should not forgot about the highest used sequence number. This is a bug in the `InmemJournal`.

Where did you notice this problem originally? Which journal did you use?

patriknw replied:
I see that you already said LeveldbJournal. That is correct behavior. The problem is in `InmemJournal`.

erinmez replied:
Hi there, the test was a means to replicate the behavior that I have encountered when a LevelDBJournal is completely gone (see above for background info). I guess the proper 1:1 test would be to shutdown the ActorSystem, delete the Journal from disk and start up again. I'm happy to rewrite this if the proposed fix actually makes sense.

patriknw replied:
ah, sorry for not reading your description fully

ok, it doesn't hurt use the highest from snapshot

erinmez committed:
Use LevelDBJournal in Test Case

The test case now properly simulates a recovery from a snapshot without
a journal.

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

erinmez commented:
CLA is signed

erinmez committed:
bump for CLA detection

erinmez committed:
added super.atStartup() & better test case description

erinmez commented:
Hi @patriknw 

I have converted the test case into a proper one that "models" what we actually encountered in our environment (snapshots but no leveldb journal artefacts).

Cheers,
Tanju

patriknw reviewed (approved):
LGTM, I can forward port this to master and then I will also fix the issue in InmemJournal and make sure we have the same for EventSourcedBehavior.

helena commented:
OK TO TEST

helena reviewed (approved):
Nice addition, LGTM

erinmez committed:
scalafmtCheck issues addressed

erinmez commented:
OK TO TEST

erinmez committed:
add return type sig to receiveCommand

erinmez commented:
@helena @patriknw this build seems to fail due to a git problem. Could the fix in #27756 be ported to release-2.5 perhaps?

raboof commented:
PLS BUILD
